File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        res = 0

        def find(root,maxTillHere):
            if not root:
                return
            nonlocal res
            if root.val >= maxTillHere:
                res+=1
                maxTillHere = max(maxTillHere,root.val)
            find(root.left,maxTillHere)
            find(root.right,maxTillHere)
        find(root,-101)
        return res
            
        

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {TreeNode}
     */
    lowestCommonAncestor(root, p, q) {
        function findAncestor(node) {
            if (p.val === node.val || q.val === node.val) return node;
            if ((p.val < node.val && node.val < q.val) || (q.val < node.val && node.val < p.val))
                return node;
            if(node.val > p.val) return findAncestor(node.left);
            if(node.val < p.val) return findAncestor(node.right);

        }

        return findAncestor(root);
    }
}
